### DISCUSSION TITLE
Requires-Dist problems when updating "Nine Rules for Writing Python Extensions in Rust"

### DISCUSSION THREAD

**Discussion by CarlKCarlK** [PyO3/maturin#1588] [Q&A] [answered]
Background: I gave a [talk last week at PyData](https://seattle2023.pydata.org/cfp/talk/7N9MQA/) about creating extensions with Rust with maturin. It was based on the ["9 rules" article](https://towardsdatascience.com/nine-rules-for-writing-python-extensions-in-rust-d35ea3a4ec29) from last year, I updated versions of Maturin, etc in the [bed-reader genomics project](https://github.com/fastlmm/bed-reader) after the talk.

I just noticed, however, that the new [*.whl](https://files.pythonhosted.org/packages/7b/21/dc6ca80d9c0de3063b1b041df52bf1fd574f5c8b124bc190429323802843/bed_reader-0.2.33-cp310-none-win_amd64.whl)'s have bed_reader_0.2.33.dist-into/METADATA without "Requires-Dist".

My Cargo.toml and pyproject.toml are in the [same (top-level) folder of my project](https://github.com/fastlmm/bed-reader). The [pyproject.toml says](https://github.com/fastlmm/bed-reader/blob/master/pyproject.toml) among other things:

```
[build-system]
requires = ["maturin==0.14.17"]
build-backend = "maturin"

[tool.maturin]
# https://github.com/PyO3/pyo3/discussions/2271
cargo-extra-args = "--features extension-module"
requires-python = ">=3.7"
requires-dist = ["numpy>=1.13.3", "pandas>=0.25.1", "pooch>=1.1.1"]
```

Any idea how it lost the "requires" info? If/when we figure it out, I'll update the article (and the project) to help others.

Thanks!
--Carl

**Comment by CarlKCarlK** [comment, marked as answer]
It looks like I need to add:

```
[project]
name = "bed-reader"
dependencies = ["numpy>=1.13.3", "pandas>=0.25.1", "pooch>=1.4.0"]
```

instead of `requires-dist`. Is that right? Is `cargo-extra-args` still needed?

**Reply by messense** [reply]
Newer maturin versions supports [PEP 621](https://peps.python.org/pep-0621/) so dependencies declaration has been moved to `[project]` table in `pyproject.toml`.

`cargo-extra-args` was removed long ago, please view the migration guide here: https://maturin.rs/migration.html.

**Comment by CarlKCarlK** [comment]
Everything seems to be working now. Thanks!
